### ISSUE TITLE
Spatio-Temporal Clustering

### ISSUE THREAD

**Issue by vagechirkov** [sktime/sktime#7054]
Hey there,

I am working with collective behavioral data. My goal is to cluster agents in space (ST) and time, so as input I have the coordinates of the agents at each time point and as output I want to have the label for each agent and each timestamp of the assignment to the cluster. See the figure for a graphical representation [adopted from 1]:

<img width="527" alt="image" src="https://github.com/user-attachments/assets/d779cfcf-7cf5-4331-a5f5-439057d80401">

[Here is the repository](https://github.com/eren-ck/spatio-temporal-clustering-benchmark) for the paper that implements most of the existing algorithms [2]. [Here is the dataset](https://zenodo.org/records/5495226) for the paper.

**Is your feature request related to a problem? Please describe.**
Implementation of the ST-clustering models

**Describe the solution you'd like**
Sktime interface for ST-clustering

**Additional context**
[This](https://github.com/GISerWang/Spatio-temporal-Clustering) github page may also be relevant. 

**References**

1. Ansari, M. Y., Ahmad, A., Khan, S. S., Bhushan, G., & Mainuddin. (2020). Spatiotemporal clustering: A review. Artificial Intelligence Review, 53(4), 2381–2423. https://doi.org/10.1007/s10462-019-09736-1
2. Cakmak, E., Plank, M., Calovi, D. S., Jordan, A., & Keim, D. (2021). Spatio-temporal clustering benchmark for collective animal behavior. Proceedings of the 1st ACM SIGSPATIAL International Workshop on Animal Movement Ecology and Human Mobility, 5–8. https://doi.org/10.1145/3486637.3489487

**Comment by fkiraly** [comment]
Nice, interesting feature request!

Some remarks:

* the algorithms are simple time series clustering algorithms, no framework extension is needed. A point of confusion may be the use of "spatio-temporal clustering" being overloaded, it could also mean that each time series has, as metadata, a geolocation. However, this is geo- or 3D locations observed over time, which is "simple" time series clustering, so it fits 1:1 into the current framework.
* the extension template to use is the one for clustering, here: https://github.com/sktime/sktime/blob/main/extension_templates/clustering.py - also see the estimator extension guide https://www.sktime.net/en/latest/developer_guide/add_estimators.html
* the license of the linked repository is GPL-3. So it would imply a GPL license for the particular algorithms. Vendoring is possible, into `sktime.libs`, with that license applying to the vendor folder. Or, we can reimplement in a clean room way (not copy-paste but from the paper) to get a permissive license verison.

**Comment by fkiraly** [comment]
Btw, since I know you are at one of the physical venues of the sprint, if you join at 13 UTC (usual meet-up time) on the discord, we are happy to get you started with more concrete pointers!

**Comment by fkiraly** [comment]
FYI @eren-ck in case you are interested to contribute! Also re the license issue, if you were willing to change this to permissive (e.g., MIT, BSD-3), it would make dissemination via `sktime` easier. You will obviously credited as the first `author` in the estimator metadata (see, e.g., https://www.sktime.net/en/latest/estimator_overview.html) if we use your code directly and not a reimplementation.

**Comment by vagechirkov** [comment]
I think it would be great to have a toy dataset to illustrate spatio-temporal clustering. I wonder if it should be a part of the PR with the implementation of the ST clustering algorithm?

**Comment by fkiraly** [comment]
sure!

Though I would suggest a separate pull request for that. It might also be great to add this in the notebook for panel tasks (no.2) which also includes clustering.

We should ensure though that we do not add too large datasets to the package itself, we´re on a trajectory to try reducing the size coming from the datasets.

**Comment by fkiraly** [comment]
Yes, I believe it is still relevant, @eren-ck - much appreciated!

FYI @vagechirkov, that is correct, right?

**Comment by vagechirkov** [comment]
The ST-DBSCAN implementation in #7353 is based on code in [this](https://github.com/eren-ck/st_dbscan) repo, which is already MIT-licensed.

@fkiraly, I think it would still be nice to have an MIT license for the code associated with the paper (Cakmak et al., 2021) in case we want to integrate more st-clustering methods from that paper in sktime. Thank you @eren-ck!
